File: backend/server/src/main/java/com/groudina/ten/demo/controllers/UserController.java
package com.groudina.ten.demo.controllers;

import com.groudina.ten.demo.datasource.DbRolesRepository;
import com.groudina.ten.demo.datasource.DbUserRepository;
import com.groudina.ten.demo.dto.ChangeUserPasswordRequestDto;
import com.groudina.ten.demo.dto.NewUserWithRole;
import com.groudina.ten.demo.dto.ResponseMessage;
import com.groudina.ten.demo.models.DbUser;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.MediaType;
import org.springframework.http.ResponseEntity;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.CrossOrigin;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import javax.validation.Valid;
import java.util.stream.Collectors;
import java.util.stream.Stream;

@Controller
@CrossOrigin(origins = {"*"}, maxAge = 3600)
@RequestMapping(path = "/api/users", produces = {MediaType.APPLICATION_JSON_VALUE})
public class UserController {
    private final Logger log = LoggerFactory.getLogger(UserController.class);

    private DbUserRepository userRepository;
    private PasswordEncoder passwordEncoder;
    private DbRolesRepository rolesRepository;

    public UserController(@Autowired DbUserRepository userRepository,
                          @Autowired PasswordEncoder passwordEncoder,
                          @Autowired DbRolesRepository rolesRepository) {
        this.userRepository = userRepository;
        this.passwordEncoder = passwordEncoder;
        this.rolesRepository = rolesRepository;
    }

    @PostMapping(value = "/create")
    @PreAuthorize("hasRole('ADMIN')")
    public Mono<ResponseEntity<ResponseMessage>> createUser(@Valid @RequestBody NewUserWithRole newUser) {
        log.info("POST: /api/users/create, body: {}", newUser);
        return userRepository.findOneByEmail(newUser.getEmail())
                .flatMap(user -> Mono.just(new ResponseEntity<>(new ResponseMessage(
                        String.format("User with email '%s' already exists!", newUser.getEmail())), HttpStatus.UNPROCESSABLE_ENTITY)))
                .switchIfEmpty(Flux.fromStream(getAllRoles(newUser.getRole()))
                        .flatMap(role -> rolesRepository.findByName(role))
                        .collect(Collectors.toList())
                        .flatMap(roles -> {
                            DbUser user = DbUser.builder().email(newUser.getEmail())
                                    .password(passwordEncoder.encode(newUser.getPassword())).roles(roles).build();

                            return userRepository.save(user)
                                    .thenReturn(ResponseEntity.ok(new ResponseMessage("User created successfully!")));
                        }));
    }

    @PostMapping(value = "/change_pwd")
    @PreAuthorize("hasRole('ADMIN')")
    public Mono<ResponseEntity<ResponseMessage>> changeUserPwd(@Valid @RequestBody ChangeUserPasswordRequestDto dto) {
        log.info("POST: /api/users/change_pwd, body: {}", dto);
        return userRepository.findOneByEmail(dto.getUserEmail())
                .flatMap(user -> {
                    user.setPassword(passwordEncoder.encode(dto.getNewPassword()));

                    return userRepository.save(user).thenReturn(ResponseEntity.ok(new ResponseMessage("Changed password successfully")));
                })
                .switchIfEmpty(Mono.just(ResponseEntity.ok(new ResponseMessage("No such user"))));
    }

    private Stream<String> getAllRoles(String topRoleName) {
        switch (topRoleName) {
            case "ROLE_ADMIN":
                return Stream.of("ROLE_ADMIN", "ROLE_TEACHER", "ROLE_STUDENT");
            case "ROLE_TEACHER":
                return Stream.of("ROLE_TEACHER", "ROLE_STUDENT");
            default:
                return Stream.of("ROLE_STUDENT");
        }
    }
}

File: backend/app/web/src/main/java/ru/selemilka/game/web/controller/UserController.kt
package ru.selemilka.game.web.controller

import org.springframework.http.MediaType
import org.springframework.http.ResponseEntity
import org.springframework.stereotype.Controller
import org.springframework.web.bind.annotation.CrossOrigin
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import ru.selemilka.game.user.postgres.model.DbUser
import ru.selemilka.game.user.postgres.repository.DbUserRepository

@Controller
@CrossOrigin(origins = ["*"], maxAge = 3600)
@RequestMapping(path = ["/user/{id}"], produces = [MediaType.APPLICATION_JSON_VALUE])
class UserController(
    // надо сделать так, чтобы из web нельзя было дёргать методы репозиториев напрямую
    // кажется, для этого нужно в мавене пометить зависимость как runtime
    val dbUserRepository: DbUserRepository,
) {
    @RequestMapping("/create/{email}/{role}")
    suspend fun create(
        @PathVariable id: Long,
        @PathVariable email: String,
        @PathVariable role: String,
    ) {
        val user = DbUser(id, email, role)
        dbUserRepository.save(user)
    }

    @RequestMapping("/find")
    suspend fun find(
        @PathVariable id: Long,
    ): ResponseEntity<DbUser> {
        val user = dbUserRepository.findById(id)
        return if (user != null) {
            ResponseEntity.ok(user)
        } else {
            ResponseEntity.notFound().build()
        }
    }
}
